src/java/org/jivesoftware/openfire/pubsub/PubSubEngine.java

```diff
@@ -1084,13 +1105,83 @@ private void getPublishedItems(PubSubService service, IQ iq, Element itemsElemen
+        // Call createNodeHelper and get the node creation status.
+        CreateNodeResponse response = createNodeHelper(service, iq, childElement, createElement);
+        if(response.getNewNode() == null) {
+            // New node creation failed
+            sendErrorPacket(iq, response.getStatus(), response.getPubsubError());
+        } else {
+            IQ reply = IQ.createResultIQ(iq);
+            Node newNode = response.getNewNode();
+            String nodeID = createElement.attributeValue("node");
+            // Include new nodeID if it has changed from the original nodeID
+            if (!newNode.getNodeID().equals(nodeID)) {
+                Element elem = reply.setChildElement("pubsub", "http://jabber.org/protocol/pubsub");
+                elem.addElement("create").addAttribute("node", newNode.getNodeID());
+            }
+            router.route(reply);
+        }
+    }
+
+    /**
+     * Response Object returned by createNodeHelper method
+     */
+    private class CreateNodeResponse {
+        private PacketError.Condition creationStatus;
+        private Node newNode;
+        private Element pubsubError;
+
+        // Getter functions
+        public PacketError.Condition getStatus() {
+            return creationStatus;
+        }
+
+        public Node getNewNode() {
+            return newNode;
+        }
+
+        public Element getPubsubError() {
+            return pubsubError;
+        }
+
+        // Setter function
+        public void setStatus(PacketError.Condition status){
+            creationStatus = status;
+        }
+
+        public void setNewNode(Node node) {
+            newNode = node;
+        }
+
+        public void setPubsubError(Element element) {
+            pubsubError = element;
+        }
+    }
+
+    /**
+     * Checks if the following conditions are satisfied and creates a node
+     * - Requester can create nodes
+     * - Instant node creation is enabled
+     * - Node does not already exist
+     * - New node configuration is valid
+     *
+     * NOTE: This method should not reply to the client
+     *
+     * @param service
+     * @param iq
+     * @param childElement
+     * @param createElement
+     * @return
+     */
+    private CreateNodeResponse createNodeHelper(PubSubService service, IQ iq, Element childElement, Element createElement) {
         // Get sender of the IQ packet
         JID from = iq.getFrom();
+        CreateNodeResponse response = new CreateNodeResponse();
         // Verify that sender has permissions to create nodes
         if (!service.canCreateNode(from) || (!UserManager.getInstance().isRegisteredUser(from) && !isComponent(from)) ) {
             // The user is not allowed to create nodes so return an error
-            sendErrorPacket(iq, PacketError.Condition.forbidden, null);
-            return;
+            response.setStatus(PacketError.Condition.forbidden);
+            return response;
```

Comment:
My previous comment would mean this would be:

return new CreateNodeResponse(PacketError.Condition.forbidden);